netket/exact.py

```diff
@@ -68,6 +69,13 @@ def lanczos_ed(
     actual_scipy_args["k"] = k
     actual_scipy_args["return_eigenvectors"] = compute_eigenvectors
 
+    if isinstance(operator, _DiscreteJaxOperator):
+        # Use the matrix free code path, as jax sparse arrays are not compatible
+        # with scipy eigsh.
+        # The default DiscreteOperator to_linear_operator implementation
+        # internally uses the sparse matrix matmul anyway.
+        matrix_free = True
+
     result = eigsh(
         operator.to_linear_operator() if matrix_free else operator.to_sparse(),
         **actual_scipy_args,
```

Comment:
can't we use `operator.to_sparse()` here, instead of complicating the logic?

I don't even know why we have this `matrix_free` path because `eigsh` operates on matrices anyway, not on lazy operators.
I suspect this is a remnant of netket v2 where we were using actual lanczos (@gcarleo do you confirm?)

-- 
I would change this line to 
```python
  operator.to_sparse(),
````
and that's it.
As for the fact that jax operators return a jax sparse matrix from `to_sparse()` maybe that was a wrong call on my part, and we should ensure that `to_sparse()` returns a scipy sparse array and `to_jax_sparse` should return a jax sparse matrix?

Alternatively, we can just add a check for jax sparse matrices and converting them to scipy if needed.

Reply:
I am not against simplicity, but converting to sparse matrix takes a lot more memory than having a matrix*vector implemented on the fly (that we don't have...) . I would not do two separate methods though: I would only keep `to_sparse` and return the appropriate type (i.e. numpy or jax, depending on the operator?) or is it problematic?

Reply:
this is scipy.sparse.linalg.eigsh not the numpy one, so it does operate on LinearOperators.
If you pass it a sparse matrix [it will convert it to one internally](https://github.com/scipy/scipy/blob/cfe80116aaa145061246b8aec0e98248fbefb835/scipy/sparse/linalg/_interface.py#L879).

Jax arrays have no built-in conversion to scipy, so we would have to do it manually for every type, which I was not willing to do.